lib/config/config.go

```diff
@@ -362,14 +345,92 @@ func (cfg *Configuration) clean() error {
 
 	// The list of ignored devices should not contain any devices that have
 	// been manually added to the config.
-	newIgnoredDevices := []protocol.DeviceID{}
+	var newIgnoredDevices []ObservedDevice
+	ignoredDevices := make(map[protocol.DeviceID]bool)
 	for _, dev := range cfg.IgnoredDevices {
-		if !existingDevices[dev] {
+		if !existingDevices[dev.ID] {
+			ignoredDevices[dev.ID] = true
 			newIgnoredDevices = append(newIgnoredDevices, dev)
 		}
 	}
 	cfg.IgnoredDevices = newIgnoredDevices
 
+	// The list of pending devices should not contain devices that were added manually, nor should it contain
+	// ignored devices.
+
+	// Sort by time, so that in case of duplicates latest "time" is used.
+	sort.Slice(cfg.PendingDevices, func(i, j int) bool {
+		return cfg.PendingDevices[i].Time.Before(cfg.PendingDevices[j].Time)
+	})
+
+	var newPendingDevices []ObservedDevice
+nextPendingDevice:
+	for _, pendingDevice := range cfg.PendingDevices {
+		if !existingDevices[pendingDevice.ID] && !ignoredDevices[pendingDevice.ID] {
+			// Deduplicate
+			for _, existingPendingDevice := range newPendingDevices {
+				if existingPendingDevice.ID == pendingDevice.ID {
+					continue nextPendingDevice
+				}
+			}
+			newPendingDevices = append(newPendingDevices, pendingDevice)
+		}
+	}
+	cfg.PendingDevices = newPendingDevices
+
+	// The list of ignored folders should not contain any devices that have
+	// been manually added to the config.
+	var newIgnoredFolders []ObservedFolder
+nextIgnoredFolder:
+	for _, ignoredFolder := range cfg.IgnoredFolders {
+		// Non-existing device, remove...
+		if !existingDevices[ignoredFolder.Device] {
+			continue
+		}
+		if fcfg, ok := existingFolders[ignoredFolder.ID]; !ok || !fcfg.SharedWith(ignoredFolder.Device) {
+			// Deduplicate
+			for _, existingIgnoredFolder := range newIgnoredFolders {
+				if existingIgnoredFolder.ID == ignoredFolder.ID && existingIgnoredFolder.Device == ignoredFolder.Device {
+					continue nextIgnoredFolder
+				}
+			}
+			newIgnoredFolders = append(newIgnoredFolders, ignoredFolder)
+		}
+	}
+	cfg.IgnoredFolders = newIgnoredFolders
+
+	// The list of pending folders should not contain any ignored folders, nor should it contain any folders that
+	// were already added to the config manually for that device, nor folder requests from ignored or non-existing devices.
+
+	// Sort by time, so that in case of duplicates latest "time" is used.
+	sort.Slice(cfg.PendingFolders, func(i, j int) bool {
+		return cfg.PendingFolders[i].Time.Before(cfg.PendingFolders[j].Time)
+	})
+
+	var newPendingFolders []ObservedFolder
+nextPendingFolder:
+	for _, pendingFolder := range cfg.PendingFolders {
+		if existingDevices[pendingFolder.Device] && !ignoredDevices[pendingFolder.Device] {
+			// Removed ignored folders
+			for _, ignoredFolder := range cfg.IgnoredFolders {
+				if pendingFolder.ID == ignoredFolder.ID && pendingFolder.Device == ignoredFolder.Device {
+					continue nextPendingFolder
+				}
```

Comment:
Regardless of what we do there's a lot of mechanics here that makes this function very long - i'd break them out into manageable pieces in their own functions...